# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze the average `total_claim_amount` by each policy type and gender for customers who have responded "Yes" to the last marketing campaign.


3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [ ]:
import pandas as pd

df = pd.read_csv('data/marketing_customer_analysis.csv')

# Convert the column names to lowercase and replace " " with "_"
df.columns = df.columns.str.lower().str.replace(' ', '_')

# 1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.
df_filtered = df[(df['total_claim_amount'] > 1000) & (df['response'] == 'Yes')]
# df_filtered 67x26

# 2. Using the original Dataframe `df`, analyze the average `total_claim_amount` by each policy type and gender for customers who have responded "Yes" to the last marketing campaign:

# Filter the original DataFrame for customers who responded "Yes"
df_yes_response = df[df['response'] == 'Yes']

# Group by policy type and gender, then calculate the average total_claim_amount
average_claim_amount = df_yes_response.groupby(['policy_type', 'gender'])['total_claim_amount'].mean().reset_index()

# Display the result
print(average_claim_amount)

# 3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers

# Group by state and count the number of customers
customers_by_state = df.groupby('state')['customer'].count().reset_index()

# Filter the results to include only states with more than 500 customers
states_with_more_than_500_customers = customers_by_state[customers_by_state['customer'] > 500]

# Display the result
print(states_with_more_than_500_customers)

# 4. Find the maximum, minimum, and median customer lifetime value by education level and gender

# Group by education level and gender, then calculate the max, min, and median customer_lifetime_value
clv_stats = df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['max', 'min', 'median']).reset_index()

# Display the result
print(clv_stats)


      policy_type gender  total_claim_amount
0  Corporate Auto      F          433.738499
1  Corporate Auto      M          408.582459
2   Personal Auto      F          452.965929
3   Personal Auto      M          457.010178
4    Special Auto      F          453.280164
5    Special Auto      M          429.527942
        state  customer
0     Arizona      1937
1  California      3552
2      Nevada       993
3      Oregon      2909
4  Washington       888


### Conclusiones del Análisis de Clientes por Estado

1. **California**: Es el estado con el mayor número de clientes, con un total de 3552 clientes. Esto sugiere que California es un mercado clave y podría beneficiarse de estrategias de marketing específicas para mantener y aumentar la base de clientes.

2. **Oregon**: Tiene 2909 clientes, lo que lo convierte en el segundo estado con más clientes. Al igual que California, Oregon representa una parte significativa de la base de clientes y podría ser un objetivo importante para campañas de retención y adquisición.

3. **Arizona**: Con 1937 clientes, Arizona también es un mercado considerable. Las estrategias de marketing en este estado podrían enfocarse en aumentar la lealtad del cliente y expandir la base de clientes.

4. **Nevada**: Tiene 993 clientes, lo que lo coloca por debajo del umbral de 1000 clientes pero aún representa una cantidad significativa. Las campañas de marketing en Nevada podrían centrarse en aumentar la visibilidad y atraer nuevos clientes.

5. **Washington**: Con 888 clientes, Washington tiene la menor cantidad de clientes entre los estados analizados. Este estado podría beneficiarse de estrategias de marketing más agresivas para aumentar la base de clientes y mejorar la retención.

En resumen, California y Oregon son los estados con el mayor número de clientes y deberían ser el foco principal de las estrategias de marketing. Arizona y Nevada también representan mercados importantes, mientras que Washington podría necesitar más atención para aumentar su base de clientes.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here